hyperapp.d.ts

```diff
@@ -36,96 +41,91 @@ export interface Component<Props> {
 export function h<Props>(
   tag: Component<Props> | string,
   props?: Props,
-  children?: VNodeChild<{} | null>[] | VNodeChild<{} | null> | number
-): VNode<Props>
+  children?: VNodeChildren
+): VNode<Props>;
 
 /** @namespace [App] */
 
-export interface State {}
-
 export interface Thunk {
-  (update: Function): any | null | void
+  (update: Function): any | null | void;
 }
 
 export type ActionResult<State extends Hyperapp.State> =
   | Partial<State>
   | Thunk
   | {}
   | null
-  | void
+  | void;
 
-export type ActionCallers<
-  State extends Hyperapp.State,
-  Actions extends Hyperapp.Actions<State>
-> = {
-  [P in keyof Actions]: (
-    data: any
-  ) => ActionResult<State> | ActionCallers<State, Actions>
+export interface Actions {
+  [action: string]: Actions | ((data: any) => ActionResult<State>);
 }
 
-export interface Actions<State extends Hyperapp.State> {
-  [action: string]:
-    | Actions<State>
-    | ((
-        state: State,
-        actions: ActionCallers<State, Actions<State>>,
-        data: any
-      ) => ActionResult<State>)
-}
+// this is done to allow ActionsImpl to reference State[P]
+export type State = { [S in keyof Actions]?: any };
```

Comment:
This is very questionable, but it seems to do the trick, anyone has any better idea?

Reply:
Can you explain a bit more ?

Reply:
You should prefer `{}` to `any`. I'm not sure `State` here is the right name. You're getting the union type of Action names.

Reply:
@Swizz When we declare ActionsImpl:
```typescript
export type ActionsImpl<
  State extends Hyperapp.State,
  Actions extends Hyperapp.Actions
> = {
  [P in (keyof Actions)]:
    | ((state: State[P], actions: Actions[P], data: any) => ActionResult<State>)
    | ActionsImpl<State, Actions>
};
```
We can do `state: State[P]` because we declared the state this way, otherwise this line would give an error.

And you may have properties in the state and not in the actions because when declaring state, you extends `Hyperapp.State<Actions>`.